### PR TITLE
Mutexes and condition variables with names

### PR DESCRIPTION
**Pull request by jessymilare**
Related to issue 372

Now `make-mutex` and `make-condition` accept an optional
argument name, which must be a symbol. When the mutex or
condition is printed, the name is also printed. The name
is immutable and can be accessed by mutex-name and
condition-name.

https://github.com/cisco/ChezScheme/issues/372

### CODE REVIEW

**Comment by jessymilare** [comment]
Added entries to LOG and release notes and updated CSUG.

File: LOG [hunk 1]
```diff
@@ -1042,3 +1042,9 @@
 - clarified required use of scheme-start to start an application
   packaged as a boot file and added a short "myecho" example.
     use.stex
+- add name fields for mutexes and condition variables, now `make-mutex` and
+  `make-condition` accept an optional argument `name`, which must be a
+  symbol.  The name is printed every time the object is printed, which is
+  usefull for debugging
```
Comment (burgerrg) [inline LOG]: usefull => useful

File: csug/threads.stex [hunk 2]
```diff
@@ -123,10 +123,16 @@ in all threads.
...
 \noskipentryheader
 \formdef{make-mutex}{\categoryprocedure}{(make-mutex)}
+\formdef{make-mutex}{\categoryprocedure}{(make-mutex \var{name})}
 \returns a new mutex object
 \listlibraries
 \endnoskipentryheader
 
+\noindent
+\var{name}, if supplied, must be a symbol which identifies the mutex. The
+name is printed every time the mutex is printed, which is usefull for
```
Comment (burgerrg) [inline csug/threads.stex]: useful

File: csug/threads.stex [hunk 3]
```diff
@@ -207,10 +213,16 @@ Using \scheme{with-mutex} is generally more convenient and safer than using
...
 \formdef{make-condition}{\categoryprocedure}{(make-condition)}
+\formdef{make-condition}{\categoryprocedure}{(make-condition \var{name})}
 \returns a new condition object
 \listlibraries
 \endnoskipentryheader
 
+\noindent
+\var{name}, if supplied, must be a symbol which identifies the condition
+object. The name is printed every time the condition is printed, which is
+usefull for debugging.
```
Comment (burgerrg) [inline csug/threads.stex]: useful

File: release_notes/release_notes.stex [hunk 4]
```diff
@@ -58,6 +58,13 @@ Online versions of both books can be found at
 %-----------------------------------------------------------------------------
 \section{Functionality Changes}\label{section:functionality}
 
+\subsection{Mutexes and condition variables can have names (9.5.1)}
+
+The procedures \scheme{make-mutex} and \scheme{make-condition} now
+accept an optional argument \scheme{name}, which must be a symbol
+that identifies the object. The name is printed every time the mutex
+or condition object is printed, which is usefull for debugging.
```
Comment (burgerrg) [inline release_notes/release_notes.stex]: useful

File: s/prims.ss [hunk 5]
```diff
@@ -1529,15 +1533,28 @@
...
-      (mutex-guardian m)
-      m)))
+  (case-lambda
+    [()
+     (let ([m ($make-mutex (mm) #f)])
+       (mutex-guardian m)
+       m)]
+    [(name)
+     (unless (symbol? name)
+      ($oops 'make-condition "~s is not a symbol" name))
```
Comment (burgerrg) [inline s/prims.ss]: should be indented 1 more space

File: s/prims.ss [hunk 6]
```diff
@@ -1561,15 +1578,28 @@
...
-      (condition-guardian c)
-      c)))
+  (case-lambda
+    [()
+     (let ([c ($make-condition (mc) #f)])
+       (condition-guardian c)
+       c)]
+    [(name)
+     (unless (symbol? name)
+      ($oops 'make-condition "~s is not a symbol" name))
```
Comment (burgerrg) [inline s/prims.ss]: should be indented 1 more space

**Comment by burgerrg** [comment]
What do you think of allowing #f for the optional second argument of make-mutex and make-condition? This would make the definitions of the single-argument cases simpler and make it easier to write functions that take an optional name.

**Comment by dybvig** [comment]
That's probably a good idea.  I don't see how it simplifies the single-argument cases, however.  In fact, they would need to explicitly allow `#f` as well as symbols, and the error messages would have to be adjusted.

**Comment by dybvig** [comment]
@jessymilare, if you use `set-who!` rather than set! for `make-mutex`, `mutex-name`, `make-condition`, and `condition-name`, you can replace the quoted names with `who` in the error messages and just put each of the tests on one line to tighten up the code.  I've been doing that with most of the things I've added recently.

**Comment by burgerrg** [comment]
For the single argument cases, I meant that one could simply call the two-argument case with #f and wouldn't need to duplicate the body.

**Comment by dybvig** [comment]
Ah, I guess you mean the zero-argument cases.  Yes, but that would send the zero-argument case through an unnecessary check.  Although if one the zero-argument case calls the one-argument case, the `case-lambda` form should be wrapped in a rec, e.g., `(rec make-mutex <case-lambda form>)` to make the call local, in which case cp0 would presumably inline the one-argument code into the zero-argument code and discard the check.  Though I would probably instead just add a helper that both cases use to make clear the test isn't necessary in the zero-argument case.

**Comment by burgerrg** [comment]
Thanks, I got my argument counts off by one! Yes, a helper without checks that is used by both front doors would be a good way to implement it.

**Comment by jessymilare** [comment]
Sorry for the polution in the commits. I use aggressive-indent-mode and I forgot to disable it before editing.

**Comment by burgerrg** [comment]
The build system found an error while building the threaded versions of Chez Scheme: [Details](https://travis-ci.org/cisco/ChezScheme/builds/510617020?utm_source=github_status&utm_medium=notification)
<pre>Warning: undeclared variable assignment to $make-mutex at line 1485, char 1 of prims.ss</pre>

**Comment by jessymilare** [comment]
> The build system found an error while building the threaded versions of Chez Scheme: [Details](https://travis-ci.org/cisco/ChezScheme/builds/510617020?utm_source=github_status&utm_medium=notification)
> 
> Warning: undeclared variable assignment to $make-mutex at line 1485, char 1 of prims.ss

Fixed.

**Comment by burgerrg** [comment]
Thank you for this pull request!
